src/Microsoft.DotNet.Tools.Builder/CompileContext.cs

```diff
@@ -0,0 +1,248 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.Tools.Compiler;
+using Microsoft.DotNet.ProjectModel.Utilities;
+
+
+namespace Microsoft.DotNet.Tools.Build
+{
+    // Knows how to orchestrate compilation for a ProjectContext
+    // Collects icnremental safety checks and transitively compiles a project context
+    internal class CompileContext
+    {
+        private readonly ProjectContext _rootProject;
+        private readonly BuilderCommandApp _args;
+        private readonly Dictionary<string, ProjectDescription> _dependencies;
+        private readonly string _outputPath;
+        private readonly string _intermediateOutputPath;
+        private readonly IncrementalPreconditions _preconditions;
+
+        public bool IsSafeForIncrementalCompilation => _preconditions.PreconditionsDetected();
+
+        public CompileContext(ProjectContext rootProject, BuilderCommandApp args)
+        {
+            _rootProject = rootProject;
+            _args = args;
+
+            // Set up Output Paths. They are unique per each CompileContext
+            // Todo: clone args and mutate the clone so the rest of this class does not have special treatment for output paths
+            _outputPath = _rootProject.GetOutputPath(_args.ConfigValue, _args.OutputValue);
+            _intermediateOutputPath = _rootProject.GetIntermediateOutputPath(_args.ConfigValue, _args.IntermediateValue, _args.OutputValue);
+
+            // Set up dependencies
+            _dependencies = GetProjectDependenciesWithSources(_rootProject, _args.ConfigValue);
+
+            //gather preconditions
+            _preconditions = GatherIncrementalPreconditions();
+        }
+
+        public bool Compile(bool incremental)
+        {
+            CreateOutputDirectories();
+
+            //compile dependencies
+            foreach (var dependency in Sort(_dependencies))
+            {
+                if (!InvokeCompileOnDependency(dependency, _outputPath, _intermediateOutputPath))
+                    return false;
+            }
+
+            //compile project
+            var success = InvokeCompileOnRootProject(_outputPath, _intermediateOutputPath);
+
+            PrintSummary(success);
+
+            return success;
+        }
+
+        private void PrintSummary(bool success)
+        {
+            if (success)
```

Comment:
Would you mind adding a comment that this code is here because it is actually the builders responsibility but it can't be added there because this code knows the compilation time.